### DISCUSSION TITLE
Adding Time in Minutes to a DateTime in Javascript Hook

### DISCUSSION THREAD

**Discussion by MikeCLX** [pocketbase/pocketbase#5312] [Q&A] [answered]
Im trying to check if a created booking will clash with an existing booking, but im having some issues with adding to a datetime to get the endTime of the booking. I have been trying to have a single source of truth in My DB but for simplicity of Clash detection i need an end time so i decided to add duration to startTime and update the end date on creating to ensure consistency  

this is my code Both newDateTime and endTime lines fail:

```
onRecordBeforeCreateRequest((e) => {
   const newBooking = e.record;
    const area = newBooking.get("area");
	 console.log(area)
    const startTime = newBooking.getDateTime("startTime");
	 console.log(startTime)
    const duration = newBooking.get("duration");
	 console.log(duration)
	 
	 
	 const newDatetime = new Date(startTime.getTime() + 5 * 60000); // 60000 milliseconds in a minute
console.log(newDatetime)	 
	 
    const endTime = startTime.time.add(duration * 60000); // Assuming duration is in minutes
console.log(endTime)
//	newBooking.set("endTime", endTime);


    // Query existing bookings in the same area
//    const existingBookings = $app.dao().findRecordsByFilter("bookings", (q) => {
//        q.where("area", area)
//         .where("startTime", "<", endTime)
//         .where("endTime", ">", startTime);
//    });

//    if (existingBookings.length > 0) {
//        throw new Error("Booking conflict detected in the same area.");
//    }
 console.log(startTime)
 console.log(endTime)


}, "bookings");
```

in the console i get as far as the duration which is 60
startTime comes back as 2024-08-05 19:00:00.000Z

I have searched and searched and tried lots of different ways of getting the correct endTime but to no avail

**Comment by ganigeorgiev** [comment, marked as answer]
>  I have been trying to have a single source of truth in My DB but for simplicity of Clash detection i need an end time so i decided to add duration to startTime and update the end date on creating to ensure consistency

I don't understand this but there are several problems with the above code sample. If you have used ChatGPT or similar to generate the code, please avoid relying on it and instead refer to the actual documentation, however scarce it is (_you can also get autocomplete for the existing methods if you specify at the top of your pb_hook files the builtin ambient TS declarations - https://pocketbase.io/docs/js-overview/#typescript-declarations-and-code-completion_).

- The provided arguments for `$app.dao().findRecordsByFilter` are wrong. Please check https://pocketbase.io/docs/js-records/#fetch-multiple-records. 

- Attaching multiple `.where` expressions to a single query is wrong because each `.where` will overwrite the previous one. If you want `AND` type of concatenation then you can use `.andWhere`. The expressions `where("area", area)` and `.where("startTime", "<", endTime)` are also wrong. For more details about the expected format please check https://pocketbase.io/docs/js-database/#where-andwhere-orwhere.

-  The Go `DateTime` is different from the standard JS `Date` object and there is no method `DateTime.getTime()` - https://pocketbase.io/jsvm/interfaces/types.DateTime.html. With the next release there are several helpers added to `DateTime` to make it easier to work with `date` type fields (e.g. `.add` and `.addDate`) but for now if you want to use regular JS `Date` objects you can do something like:
    ```js
    // convert from rfc3339 to iso8601 so that we can create a JS date with it
    const startTimeStr = newBooking.getString("startTime").replace(" ", "T")

    // add minutes duration to startTime
    const endTime = new Date(startTimeStr)
    endTime.setMinutes(endTime.getMinutes() + duration)

    // the setter should cast it automatically to the underlying `date` field type value
    newBooking.set("endTime", endTime)
    ```

**Reply by MikeCLX** [reply]
Thanks so much, I had not got as far as the query as i was having so much issue with the time :)
In the mean time I had worked through the documentation and got the query do the right thing so the entire function now does exactly what i was hoping for.

Regards,

Mike
